sources/growpart/src/diskpart.rs

```diff
@@ -140,3 +148,63 @@ impl DiskPart {
         Ok(gpt)
     }
 }
+
+struct WatchPart {
+    inotify: Inotify,
+    filename: PathBuf,
+}
+
+impl WatchPart {
+    /// Given a path to a partition, set up an inotify watch that will record
+    /// create and delete events.
+    fn new<P>(path: P) -> Result<Self>
+    where
+        P: AsRef<Path>,
+    {
+        let path = path.as_ref();
+        let dirname = path.parent().context(error::FindParentDirectory { path })?;
+
+       // When the kernel reloads the partition table, we expect two events, when udev deletes and
+       // then recreates the path. This isn't synchronized with our code, so to avoid races we need
+       // to watch for both events.
+        let mut inotify = Inotify::init().context(error::InitInotify)?;
+        inotify
+            .add_watch(&dirname, WatchMask::CREATE | WatchMask::DELETE)
+            .context(error::AddInotifyWatch)?;
+
+        let filename = path.file_name().context(error::FindFileName { path })?;
+        let filename = Path::new(filename).to_path_buf();
+
+        Ok(WatchPart { inotify, filename })
+    }
+
+    /// Poll the inotify watch until the create and delete events are found.
+    fn wait(&mut self) -> Result<()> {
+        let mut need_create = true;
+        let mut need_delete = true;
+        let mut buf = [0; 1024];
+
+        while need_create || need_delete {
+            let events = self
+                .inotify
+                .read_events_blocking(&mut buf)
+                .context(error::ReadInotifyEvents)?;
+
+            for event in events {
+                if let Some(event_file) = event.name {
+                    if self.filename != Path::new(event_file) {
+                        continue;
+                    } else if event.mask == EventMask::DELETE {
```

Comment:
```suggestion
                        continue;
                    }

                    if event.mask == EventMask::DELETE {
```
nit: these are logically separate conditionals; this would have to run the same way since the first one breaks out.

Reply:
True, and I had it the other way at first, but I thought this was slightly more readable since there are really only these three branches to consider.